### DISCUSSION TITLE
Detecting last column despite missing vertical border

### DISCUSSION THREAD

**Discussion by nicopicchio** [jsvine/pdfplumber#1223] [Ask for help with specific PDFs]
I am trying to get data from a PDF table that unfortunately has a missing border on the right side of the page as shown in the image below. This means that with my current code I cannot get the data from the last column.

![image](https://github.com/user-attachments/assets/f657964f-c506-4cb4-adb0-4f42990950f8)

I think there might be a way to only select the rows by changing the `table_settings` parameter of the `extract_table()` method but I can't for the life of me figure out how.

Currently my code looks as follows:
```python
import pdfplumber

with pdfplumber.open("pdfs/pdf_file.pdf") as pdf:
    p1 = pdf.pages[1]
    table_settings = {
        "horizontal_strategy": "text"
    }
    tables = p1.extract_tables(table_settings)
    print(tables)
```

Would someone please be able to point me in the right direction?

Thank you.

**Comment by bloktom** [comment]
I have the same issue:

![image](https://github.com/user-attachments/assets/11443931-b50b-4c2d-89e9-6feeaba4959f)

It is possible to use "vertical_strategy" : "text" as well but this will extract too many values for me.

@jsvine do you know more about this?

**Reply by nicopicchio** [reply]
yeah, I tried `"vertical_strategy": "text"` last night and I could achieve what I wanted to do but only limiting the script to the first page where there is a complete table (apart from that missing right border). Other pages have tables that span across multiple pages that are without bottom border on one page and top border on the following so that messes up my implementation.

I have not tried to combine the two settings but I will give it a go tonight.

**Comment by cmdlineluser** [comment]
You can also pass your own vertical lines with `explicit_vertical_lines`

The simplest strategy may be to use the position of the right-most char on the page:

```ipython
>>> max(page.chars, key=lambda char: char['x1'])['x1'] + 3
562.2739981415099
```

```python
page.extract_table({
    "explicit_vertical_lines": [max(page.chars, key=lambda char: char['x1'])['x1'] + 3]
})
```

If there are multiple tables a more robust approach is probably to use `.find_tables()` which returns Table objects.

```ipython
>>> page.find_table()
<pdfplumber.table.Table at 0x12d6ccb20>
>>> page.find_table().bbox
(55.96562783864786, 94.85909107729572, 556.4055217861666, 227.83481615943583)
```

This lets you use the table top/bottom lines to first crop out a page area and search only within that area for a specific marker.

```python
for table in page.find_tables():
    _, top, _, bottom = table.bbox
    crop = page.crop((page.bbox[0], top, page.bbox[2], bottom))
    line = max(crop.chars, ...)['x1'] + 3
    crop.extract_table({"explicit_vertical_lines": [line]})
```
